src/util/threadpool.rs

```diff
@@ -0,0 +1,601 @@
+    }
+
+    // `on_task_finished` is called when the thread finished a task.
+    // It will clean up the remaining information of the task.
+    fn on_task_finished(&mut self, group_id: &T) {
+        self.tasks.on_task_finished(group_id);
+    }
+
+    fn on_task_started(&mut self, group_id: &T) {
+        self.tasks.on_task_started(group_id);
+    }
+
+    fn pop_task(&mut self) -> Option<Task<T>> {
+        self.tasks.pop()
+    }
+
+    #[inline]
+    fn stop(&mut self) {
+        self.stop = true;
+    }
+
+    #[inline]
+    fn is_stopped(&self) -> bool {
+        self.stop
+    }
+}
+
+/// `ThreadPool` is used to execute tasks in parallel.
+/// Each task would be pushed into the pool, and when a thread
+/// is ready to process a task, it will get a task from the pool
+/// according to the `ScheduleQueue` provided in initialization.
+pub struct ThreadPool<Q, T> {
+    task_pool: Arc<(Mutex<TaskPool<Q, T>>, Condvar)>,
+    threads: Vec<JoinHandle<()>>,
+    task_count: Arc<AtomicUsize>,
+}
+
+impl<Q: ScheduleQueue<T> + Send + 'static, T: Hash + Eq + Send + Clone + 'static> ThreadPool<Q, T> {
+    pub fn new(name: String, num_threads: usize, queue: Q) -> ThreadPool<Q, T> {
+        assert!(num_threads >= 1);
+        let task_pool = Arc::new((Mutex::new(TaskPool::new(queue)), Condvar::new()));
+        let mut threads = Vec::with_capacity(num_threads);
+        let task_count = Arc::new(AtomicUsize::new(0));
+        // Threadpool threads
+        for _ in 0..num_threads {
+            let tasks = task_pool.clone();
+            let task_num = task_count.clone();
+            let thread = Builder::new()
+                .name(name.clone())
+                .spawn(move || {
+                    let mut worker = Worker::new(tasks, task_num);
+                    worker.run();
+                })
+                .unwrap();
+            threads.push(thread);
+        }
+
+        ThreadPool {
+            task_pool: task_pool,
+            threads: threads,
+            task_count: task_count,
+        }
+    }
+
+
+    pub fn execute<F>(&mut self, group_id: T, job: F)
+        where F: FnOnce() + Send + 'static
+    {
+        self.task_count.fetch_add(1, AtomicOrdering::SeqCst);
+        let &(ref lock, ref cvar) = &*self.task_pool;
+        let mut meta = lock.lock().unwrap();
+        meta.push_task(group_id, job);
+        cvar.notify_one();
+    }
+
+    #[inline]
+    pub fn get_task_count(&self) -> usize {
+        self.task_count.load(AtomicOrdering::SeqCst)
+    }
+
+    pub fn stop(&mut self) -> Result<(), String> {
+        {
+            let &(ref lock, ref cvar) = &*self.task_pool;
+            let mut tasks = lock.lock().unwrap();
+            tasks.stop();
+            cvar.notify_all();
+        }
+        let mut err_msg = String::new();
+        while let Some(t) = self.threads.pop() {
+            if let Err(e) = t.join() {
+                write!(&mut err_msg, "failed to join thread with err :{:?}", e).unwrap();
```

Comment:
no separator for the multi-error messages in the string? 

If we have two errors, the message may like "failed to join thread... failed to join thread"